Add documentation for project-owned variable sets (#790)
hashicorp/terraform-docs-common: website/docs/cloud-docs/workspaces/variables/index.mdx

| Project-Scoped Variable Set   | Automatically applied to all current and future workspaces within a project.       | [Create Variable Sets](/terraform/cloud-docs/workspaces/variables/managing-variables#variable-sets) and [Variable Sets API](/terraform/cloud-docs/api-docs/variable-sets)                                                                                                                                                                              |
| Global Variable Set           | Automatically applied to all current and future workspaces within an organization. | [Create Variable Sets](/terraform/cloud-docs/workspaces/variables/managing-variables#variable-sets) and [Variable Sets API](/terraform/cloud-docs/api-docs/variable-sets)                                                                                                                                                                              |

## Variable Set Ownership

> Variable set ownership determines whether the variable set is managed at the organization level or the project level.

Reviewer: Suggestion to make this a little simpler:
```suggestion
Projects and organizations can both own variable sets. The owner of a variable set can determine the precedence of that set.
```

> Use organization-owned variable sets to share variables across multiple projects. Because they can be applied to workspaces in any project in the organization, managing organization-owned variable sets requires a higher level of permissions.

Reviewer: Love the clear recommendation here! Suggestion for flow and active voice:
```suggestion
Use organization-owned variable sets to share variables across multiple projects. Managing organization-owned variable sets [requires a higher permission level](/terraform/cloud-docs/workspaces/variables/managing-variables#permissions) because you can apply these sets to any project in your organization.
```

> Use project-owned variable sets to share variables across multiple workspaces within a single project. Project-owned variable sets allow for more localized management because they require permissions on the project rather than on the entire organization.

Reviewer: ```suggestion
Use project-owned variable sets to share variables across multiple workspaces within a single project. Project-owned variable sets only require permissions on the project itself, rather than organization-level permissions.
```

> 

Reviewer: Right above here in the precedence section can you add information about how the entity owning the variable set affects precedence? With an example too please!

If [prioritized](/terraform/cloud-docs/workspaces/variables#precedence-with-priority-variable-sets), variables in a global variable set have precedence over all other variables with the same key.

> ### 2. Priority project-scoped, organization-owned variable sets

Reviewer: I will admit I totally did not understand what this meant at first! I think we should really spell it out for clarity:
```suggestion
### 2. Priority organization-owned variable set scoped to a project
```

> ### 2. Priority organization-owned variable set scoped to a project

Author: @rkoron007 I applied your suggestion to these titles, but I'm thinking now it makes more sense to have the scope part come first because the content of the sections focuses more on scope. Maybe something more like `Priority project-scoped variable set owned by an organization`.

Reviewer: You make a _really_ good point.  Though, I do kind of worry that without the extra focus on how the ownership affects precedent, the headers become harder to differentiate one from one another.  

For example, the following headers feel easier confuse:
```
2. Priority project-scoped variable set owned by an organization

...and then below...

3. Priority project-scoped variable set owned by a project
```

I always worry people don't read so I try to shift the important details to the left 😂 However, I don't feel super strongly about this, so if you think it'd be better to go by scope go for it!

> If [prioritized](/terraform/cloud-docs/workspaces/variables#precedence-with-priority-variable-sets), variables in a priority project-scoped variable set have precedence over variables with the same key set at a more specific scope. Prioritized variables sets owned by organizations take precedence over those owned by projects.

Reviewer: To really drive the point home:
```suggestion
If [prioritized](/terraform/cloud-docs/workspaces/variables#precedence-with-priority-variable-sets), variables in a priority project-scoped variable set have precedence over variables with the same key set at a more specific scope. Prioritized variables sets owned by organizations take precedence over priority variable sets owned by projects.
```

> ### 3. Priority project-scoped, project-owned variable sets

Reviewer: ```suggestion
### 3. Priority project-owned variable set scoped to a project
```

If [prioritized](/terraform/cloud-docs/workspaces/variables#precedence-with-priority-variable-sets), variables in a priority project-scoped variable set have precedence over variables with the same key set at a more specific scope.

### 4. Priority workspace-scoped, organization-owned variable sets

If [prioritized](/terraform/cloud-docs/workspaces/variables#precedence-with-priority-variable-sets), variables in a priority workspace-scoped variable set have precedence over variables with the same key set at a more specific scope. Prioritized variables sets owned by organizations take precedence over those owned by projects.

> ### 5. Priority workspace-scoped, project-owned variable sets

Reviewer: ```suggestion
### 5. Priority project-owned variable set scoped to a workspace
```

Variables sets owned by projects take precedence over those owned by organizations.

### 10. Workspace-scoped, organization-owned variable sets

Variables in workspace-scoped variable sets are only applied to a subset of workspaces in an organization.

When workspace-scoped variable sets have conflicting variables, HCP Terraform compares the variable set names and uses values from the variable set with lexical precedence. Terraform and HCP Terraform operate on UTF-8 strings, and HCP Terraform sorts variable set names based on the lexical order of Unicode code points.

For example, if you apply `A_Variable_Set` and `B_Variable_Set` to the same workspace, HCP Terraform will use any conflicting variables from `A_Variable_Set`. This is the case regardless of which variable set has been edited most recently. HCP Terraform only considers the lexical ordering of variable set names when determining precedence.

### 11. Project-scoped, project-owned variable sets

> Workspace-specific variables and workspace-scoped variable sets always take precedence over project-scoped variable sets that are applied to workspaces within a project.

Reviewer: I don't think we need this sentence because the section ordering makes it obvious that workspace-scoped variables take precedence? Curious how you feel though!

Author: That makes sense to me too! This line in particular is kind of confusing too now that I'm reading it because the section is about project scoped variables but then we start this sentence off talking about workspace scoped variables.

Reviewer: ```suggestion
```

Variables in project-scoped variable sets are only applied to the workspaces within the specified projects.

When project-scoped variable sets have conflicting variables, HCP Terraform compares the variable set names and uses values from the variable set with lexical precedence. Terraform and HCP Terraform operate on UTF-8 strings, and HCP Terraform sorts variable set names based the on lexical order of Unicode code points.

For example, if you apply `A_Variable_Set` and `B_Variable_Set` to the same project, HCP Terraform uses any conflicting variables from `A_Variable_Set`. This is the case regardless of which variable set has been edited most recently. HCP Terraform only considers the lexical ordering of variable set names when determining precedence.

Variables sets owned by projects take precedence over those owned by organizations.

### 12. Project-scoped, organization-owned variable sets

> Workspace-specific variables and workspace-scoped variable sets always take precedence over project-scoped variable sets that are applied to workspaces within a project.

Reviewer: Again, don't feel like we need this, and can you update this to mention organizations?

For example, if you apply `A_Variable_Set` and `B_Variable_Set` to the same project, HCP Terraform uses any conflicting variables from `A_Variable_Set`. This is the case regardless of which variable set has been edited most recently. HCP Terraform only considers the lexical ordering of variable set names when determining precedence.

### 13. Global variable sets

Workspace and project-scoped variable sets always take precedence over global variable sets that are applied to all workspaces within an organization. Terraform does not allow global variable sets to contain variables with the same key, so they cannot conflict.

### 14. `*.auto.tfvars` variable files

Variables in the HCP Terraform workspace and variables provided through the command line always overwrite variables with the same key from files ending in `.auto.tfvars`.

### 15. `terraform.tfvars` variable file

Variables in the `.auto.tfvars` files take precedence over variables in the `terraform.tfvars` file.

> 

Reviewer 2: Of course! Thanks for catching that.

### 5. Priority workspace-scoped variable set owned by an project

If [prioritized](/terraform/cloud-docs/workspaces/variables#precedence-with-priority-variable-sets), variables in a priority workspace-scoped variable set have precedence over variables with the same key set at a more specific scope.

### 6. Command line argument variables

When using a CLI workflow, variables applied to a run with either `-var` or `-var-file` overwrite workspace-specific and variable set variables that have the same key.

### 7. Local environment variables prefixed with `TF_VAR_`

When using a CLI workflow, local environment variables prefixed with `TF_VAR_` (e.g., `TF_VAR_replicas`) overwrite workspace-specific, variable set, and `.auto.tfvars` file variables that have the same key.

### 8. Workspace-specific variables

Workspace-specific variables always overwrite variables from variable sets that have the same key. Refer to [overwrite variables from variable sets](/terraform/cloud-docs/workspaces/variables/managing-variables#overwrite-variable-sets) for details.

### 9. Workspace-scoped variable owned by a project

> Variables in workspace-scoped variable sets are only applied to a subset of workspaces in an organization.

Reviewer: Should this be project?

When workspace-scoped variable sets have conflicting variables, HCP Terraform compares the variable set names and uses values from the variable set with lexical precedence. Terraform and HCP Terraform operate on UTF-8 strings, and HCP Terraform sorts variable set names based on the lexical order of Unicode code points.

For example, if you apply `A_Variable_Set` and `B_Variable_Set` to the same workspace, HCP Terraform will use any conflicting variables from `A_Variable_Set`. This is the case regardless of which variable set has been edited most recently. HCP Terraform only considers the lexical ordering of variable set names when determining precedence.

> Variables sets owned by projects that are scoped to workspaces take precedence over workspace-scoped sets owned by organizations.

Reviewer: I think this reads a little clearly on second read:
```suggestion
Variables sets scoped to workspaces that are owned by projects take precedence over sets with the same scope that are owned by organizations.
```

For example, variables in a priority global variable set would take precedence over all variables with the same key.

If two priority variable sets with the same scope and ownership include the same variable key, HCP Terraform will determine precedence by the alphabetical order of the variable sets' names.

While a priority variable set can enforce that Terraform variables use designated values, it does not guarantee that the configuration uses the variable. A user can still directly modify the Terraform configuration to remove usage of a variable and replace it with a hard-coded value. For stricter enforcement, we recommend using policy checks or run tasks.

## Precedence example

Consider an example workspace that has the following variables applied:

| (**Scope**) Source                                            | Region      | Var1 | Replicas |
| ------------------------------------------------------------- | ----------- | ---- | -------- |
| Priority **global** variable set                              | `us-east-1` |      |          |

> | Priority **project-scoped** organization-owned variable set   | `us-east-2` |      |          |

Reviewer: To make the differences between rows more apparent, how would you feel about moving the ownership changes to the left so they can more visible? Bringing the focus to the ownership instead of just the scope.

For example, moving organization-owned before "project-scoped":
```suggestion
| Priority organization-owned **project-scoped** variable set   | `us-east-2` |      |          |
```

Reviewer: We could also have an ownership column to really drive the point home. What do you think?

Author: I tried out having extra columns for the different factors (priority, ownership, scope) but then I though it's kind of confusing to then also have the columns for the example variables. 🤔  It looks like this:

| Source                                                        | Priority | Ownership    | Scope     | Region      | Var1 | Replicas |
| ------------------------------------------------------------- | -------- | ------------ | --------- | ----------- | ---- | -------- |
| Priority **global** variable set                              | true     | organization | global    | `us-east-1` |      |          |
| Priority organization-owned **project-scoped** variable set   | true     | organization | project   | `us-east-2` |      |          |
| Priority project-owned **project-scoped** variable set        | true     | project      | project   | `eu-east-2` |      |          |
| Priority organization-owned **workspace-scoped** variable set | true     | organization | workspace | `us-west-1` |      |          |
| Priority project-owned **workspace-scoped** variable set      | true     | project      | workspace | `eu-west-1` |      |          |
| Command line argument                                         | N/A      |              | workspace | `us-west-2` |      | `9`      |
| Local environment variable                                    | N/A      |              | workspace |             |      | `8`      |
| **Workspace-specific** variable                               | N/A      |              | workspace |             | `h`  | `1`      |
| Project-owned **workspace-scoped** variable set               | false    | project      | workspace |             | `y`  | `2`      |
| Organization-owned **workspace-scoped** variable set          | false    | organization | workspace |             | `z`  | `3`      |
| Project-owned **project-scoped** variable set                 | false    | project      | project   |             |      | `4`      |
| Organization-owned **Project-scoped** variable set            | false    | organization | project   |             |      | `5`      |
| **Global** variable set                                       | false    |              | global    |             |      | `6`      |

Reviewer: Oh geeze yeah, that's rough 😂  I think it makes sense to skip the ownership column then!

Author: Not sure if now it's too much information in this section, but I added a second, separate table with all the factors that are used to determine the precedence. I liked having the columns with the factors because I think it's easier to view the precedence order and how the factors affect it. Let me know what you think!
